### PR TITLE
fix(popover): don't clone anchor with ref if it's not using forwardRef

### LINKED ISSUE #3265: Popover is not appearing if anchor component does not forward ref
## Feature description
Popover should not be dependent on attributes of child component.

Check 2 examples in the codesandbox for better understanding
https://codesandbox.io/s/statefulpopover-56pd9?file=/example.js

### PR DESCRIPTION
**Pull request by devanshj**
Closes #3265

#### Description
We're already logging a warning if the anchor ref is not being set, but if it's a functional component without `forwardRef` React logs an error that you can't set ref and bails out thus our warning doesn't log. So the fix doesn't set the ref while cloning hence out warning would get logged.

#### Scope
Minor: New Feature
(going for a minor because it's also kinda a feature)

### CODE REVIEW

**Comment by devanshj** [comment]
Hmm the snapshots are apparently passing locally 🤔 let me try undoing them to see if it passes in the CI

File: src/popover/popover.js [hunk 1]
```diff
@@ -361,10 +361,20 @@ class Popover extends React.Component<PopoverPropsT, PopoverPrivateStateT> {
       return null;
     }
 
-    const isValidElement = React.isValidElement(anchor);
-    const anchorProps = this.getAnchorProps();
+    let anchorProps = this.getAnchorProps();
+
+    if (typeof anchor === 'object' && React.isValidElement(anchor)) {
+      if (typeof anchor.type === 'function') {
```
Comment (chasestarr) [inline src/popover/popover.js:367]: Reading undocumented values on the react element is putting this component at risk of regressions on react minor version changes. Silently not forwarding the ref is going to lead to more unexpected behavior

File: src/popover/popover.js [hunk 2]
```diff
@@ -361,10 +361,20 @@ class Popover extends React.Component<PopoverPropsT, PopoverPrivateStateT> {
...
+        /* if is non-forwardRef functional component
+         * (if it's wrapped by forwardRef then the type is no longer a function)
+         * then remove ref because rendering non-forwardRef functional component throws
+         * an error and which makes tether unable to show the ref not set warning
```
Comment (chasestarr) [inline src/popover/popover.js:371]: Please see the related issue, there is still a warning provided by react that seems actionable-enough to me

File: src/popover/popover.js (hunk 1 above)
Comment (devanshj) [reply src/popover/popover.js:367]: Yeah fair enough let me see if I can come up with something

File: src/popover/popover.js (hunk 2 above)
Comment (devanshj) [reply src/popover/popover.js:371]: Yeah I replied there

File: src/popover/popover.js (hunk 1 above)
Comment (chasestarr) [reply src/popover/popover.js:367]: 👍 Thanks for looking into the problem. I'm happy to land a solution, but hopefully in a more maintainable way

**Comment by chasestarr** [comment]
> Also how does buildkite work like what does "passed & block" mean? Coz it said that before and now that I see it had failures... Like does a maintainer has to do something and unblock it and then it actually runs?

It doesn't run CI until we have a chance to read through the diff and verify nothing suspicious is happening. Someone with commit access will need to 'unblock' CI

File: src/popover/popover.js (hunk 1 above)
Comment (devanshj) [reply src/popover/popover.js:367]: Yeah no problems 👍

**Comment by devanshj** [comment]
> It doesn't run CI until we have a chance to read through the diff and verify nothing suspicious is happening. Someone with commit access will need to 'unblock' CI

Ahhh exactly what I guessed, I used to think the "Passed & Blocked" part meant it ran and passed and when I come here after sometime boom it fails xD

File: src/layer/tether.js [hunk 3]
```diff
@@ -48,9 +48,9 @@ class Tether extends React.Component<TetherPropsT, TetherStateT> {
           if (__DEV__) {
             // eslint-disable-next-line no-console
             console.warn(
-              `[baseui][TetherBehavior] ref has not been passed to the Popper's anchor element.
-              See how to pass the ref to an anchor element in the Popover example
-              http://baseui.design/components/popover#anchor-ref-handling-example`,
+              `[baseui][TetherBehavior] ref has not been passed to the Popper's anchor element.\n` +
```
Comment (nadiia) [inline src/layer/tether.js:51]:
```suggestion
              '[baseui][TetherBehavior] ref has not been passed to the Popper's anchor element.\n' +
```

File: src/layer/tether.js [hunk 4]
```diff
@@ -48,9 +48,9 @@ class Tether extends React.Component<TetherPropsT, TetherStateT> {
...
+                `See how to pass the ref to an anchor element in the Popover example\n` +
```
Comment (nadiia) [inline src/layer/tether.js:52]:
```suggestion
                'See how to pass the ref to an anchor element in the Popover example\n' +
```

File: src/layer/tether.js [hunk 5]
```diff
@@ -48,9 +48,9 @@ class Tether extends React.Component<TetherPropsT, TetherStateT> {
...
+                `http://baseui.design/components/popover#anchor-ref-handling-example`,
```
Comment (nadiia) [inline src/layer/tether.js:53]:
```suggestion
                'http://baseui.design/components/popover#anchor-ref-handling-example',
```

**Comment by chasestarr** [comment]
Closing for now, but feel free to reopen

**Comment by devanshj** [comment]
Yeah cool! I was actually gonna send a proposal on react repo to have `isRefableElement` gaurd in `react-is` package (it already has `isForwardRef` but that's not enough coz the element could be a dom element too) so we could use that but haven't got time to send it. In case they accept the proposal I'll open a PR to use it here.

**Pull request closed without merging**